tests/phpunit/AutoReview/Event/SubscriberProvider.php

```diff
@@ -130,4 +113,24 @@ public static function subscriberSubscriptionMethodsProvider(): Generator
     {
         yield from self::provideSubscriberSubscriptionMethods();
     }
+
+    /**
+     * @return string[]
+     */
+    private static function retrieveSubscriberSubscriptions(ReflectionClass $subscriberReflection): array
+    {
+        $methods = array_filter(
+            $subscriberReflection->getMethods(ReflectionMethod::IS_PUBLIC),
+            static function (ReflectionMethod $method): bool {
+                return strpos($method->getName(), 'on') === 0;
+            }
+        );
+
+        return array_map(
+            static function (ReflectionMethod $method): string {
+                return $method->getName();
+            },
+            $methods
+        );
```

Comment:
How about if:
- Each event will declare in a constant which interface must subscribers implement.
- The loop in `dispatch()` above for each event will grab its expected subscriber interface, and check every subscriber against this interface.
- Subscribers would have a single entry point for events, where they would shuffle incoming events into handlers according to their type.
- Thus it seems we could avoid most introspection.

Another approach:
- Since our `on...` methods are typed, here you can make a list not only of methods, but also of their argument's type. Say, a map with argument type as a key. So this will be the only place to use reflection. Up above in `dispatch()` it should be possible to pattern-match incoming events into handlers.
- This method still requires reflection, but avoids adding a bunch of extra interfaces.

Reply:
I think I would prefer going with generating a map. But that brings the question on how you keep that map up to date (in a composer script when dumping the autoload for example?)

Reply:
Alright, here you go #1084

Reply:
Will try to compare performance wise; I otherwise I think both approach are fine it's just a question of flavour so up to @infection/core team to choose.

But otherwise I was more thinking of generating a static map, rather than doing so at run-time

Reply:
Performance wise it won't make much sense this way or another because this is a one-off operation.

Reply:
But if you want to look at this this way, consider that in #1084 we run Reflection only once per registration, where here you want to run it slightly simpler, but still more complex operation than just a map lookup, on each `dispatch()` call.